src/components/MediaLibrary/MediaLibrary.js

```diff
@@ -24,6 +25,60 @@ import { Icon } from 'UI';
 const IMAGE_EXTENSIONS_VIEWABLE = [ 'jpg', 'jpeg', 'webp', 'gif', 'png', 'bmp', 'tiff', 'svg' ];
 const IMAGE_EXTENSIONS = [ ...IMAGE_EXTENSIONS_VIEWABLE ];
 
+class CardImage extends React.Component {
+  state = {
+    imageURL: "",
+    isFetching: false,
+  };
+
+  loadImage() {
+    const { image, getCachedImageURLByID, cacheImageURLByID } = this.props;
+    const { imageURL: existingImageURL, isFetching } = this.state;
+
+    if (existingImageURL !== "" || isFetching) {
+      return;
+    }
+
+    if (getCachedImageURLByID && image.key) {
+      const imageURL = getCachedImageURLByID(image.key);
+      if (imageURL) {
+        this.setState({ imageURL });
+        return;
+      }
+    }
+
+    if (image.url) {
+      this.setState({ imageURL: image.url });
+      if (image.key && cacheImageURLByID) {
+        cacheImageURLByID(image.key, image.url);
+      }
+      return;
+    }
+
+    if (image.getBlobPromise) {
+      this.setState({ isFetching: true });
+      image.getBlobPromise().then(blob => {
+        const imageURL = window.URL.createObjectURL(blob);
+        this.setState({ imageURL, isFetching: false });
+        if (image.key && cacheImageURLByID) {
+          cacheImageURLByID(image.key, imageURL);
+        }
+      });
+    }
+  }
+
+  render() {
+    const { imageURL, isFetching } = this.state;
+
+    if (imageURL === "" && !isFetching) {
+      this.loadImage();
```

Comment:
We're calling `setState` during the render - we need to move this state management to Redux.